modify `ct` provider to copy selected directories and compile them

talentdeficit opened the pull request:
modify `ct` provider to copy selected directories and compile them alongside their source (ie, if `some_tests` is a directory that contains test suites beams resulting from compiling them will be
placed in `some_tests` in the appropriate `_build` directory)

also an actual test suite that tests some relevant things

and logs are now put in `_build/PROFILE/logs` by default

drawbacks to this provider:

test specifications aren't really supported. short of parsing them ourselves and deconstructing all their paths there's no way to get them to work well with the approach this provider takes (copy any test directories into `_build` then work as closely as possible to `ct_run` from there). instead, this provider just punts and doesn't copy or compile any directories leaving it up to the test specification user to figure paths out. the specification will, however, be passed to `ct:run_test/1`

similarly, cover specifications as used by ct aren't reasonable to support without reimplementing large chunks of `common_test` itself. you can pass a cover specification as an argument but unless used in conjunction with a test specification it may not work as expected

from a pragmatic standpoint, this provider now covers probably 90% of common test use cases and the other 10% are complex enough users probably wouldn't trust rebar to get things right anyways

ferd commented:
So how does that compare to the current provider in terms of features? Is this a step backwards in any sort of way? What should we expect to break or be fixed compared to the current one?

Also doesn't moving the whole directory break the assumption that everything goes into the ebin/ directory and then break `rebar3 shell` and other similar commands?

talentdeficit commented:
short answer: it's better

fixed:
* 'data_dir'
* test specifications
* test resolution (combining the 'dir' and 'suite' opts)
* emakefiles work with 'auto_compile'

better but still sort of broken:
* cover specifications work fine alongside test specifications but may still have path issues unless written specifically for use with rebar and work badly in combination with the cover provider

broken but fixable
* 'shell' doesn't include ct test directories by default as the only way to discover them is to run ct basically. there's two possible solutions to this (one is to add an option to the ct provider to do a dry run where tests are setup but not run so 'rebar3 do ct --dry, shell' works the other is to add the default 'test' paths in the shell provider and provide some way to add additional paths on top of that if necessary)
* probably other things. as you can see it's a fairly complex changeset. it's much better tested than the old provider however and it is relatively straight forward to add additional tests unlike the old provider

ferd commented:
Ah. But you did say test specifications are not supported, though you fixed them? The problem is specifying the paths, right?

talentdeficit commented:
they work but they work by basically short circuiting the provider. no directories are copied or compiled by rebar, the provider just resolves arguments (from 'ct_opts' in the config and the command line options) and then calls 'ct:run_test/1'. to support test specs properly they'd have to be interpreted and rewritten with modified paths. common test, unfortunately, makes this difficult to do with the way it exposes (or doesn't expose) it's internals

ferd commented:
Alright. I'll try to review this PR sometime this week. I'm guessing worst case would be to also copy the spec file and change the arg internally to the rewritten one, but that also sounds like a lot of trouble.

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+        %% a single dir defined, this is ok
+        {Dirs, Suites} when is_integer(hd(Dirs)), is_list(Suites) -> Opts;
+        %% still a single dir defined, adjust to make acceptable to ct
+        {[Dir], Suites} when is_integer(hd(Dir)), is_list(Suites) ->
```
What are the `is_integer()` checks for? If it's a list of characters?

talentdeficit replied:
yeah, that it's `"foo"` and not `["foo"]`

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+    end.
+
+discover_testspec(_TestSpec, Opts) ->
+    lists:keydelete(auto_compile, 1, Opts).
```
Is this function incomplete or there is actually nothing to discover in testspecs? If so, it might be worth a comment.

talentdeficit replied:
it's complete. presence of a test spec skips all the copying and compiling of suites

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+    case proplists:get_value(dir, Opts) of
+        undefined -> {error, {no_tests_specified, Opts}};
+        %% dir is a single directory
+        Dir when is_list(Dir), is_integer(hd(Dir)) ->
```
Same question with that integer check ehre

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+reduce_path(Acc, ["."|Rest])       -> reduce_path(Acc, Rest);
+reduce_path([_|Acc], [".."|Rest])  -> reduce_path(Acc, Rest);
+reduce_path([], [".."|Rest])       -> reduce_path([], Rest);
+reduce_path(Acc, [Component|Rest]) -> reduce_path([Component|Acc], Rest).
```
I'll make a mental note to remember all of that stuff in case it's needed again to be extracted into `rebar_dir` or `rebar_utils`

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+handle_results({error, Reason}) ->
+    ?PRV_ERROR({error_running_tests, Reason});
+handle_results(_) ->
+    ok.
```
I'd probably break this up in two functions -- `sum_results` --> `handle_results`. One is in charge of bundling them together, and the other to react according to the result set.

ferd commented on src/rebar_prv_common_test.erl:
```diff
@@ -37,54 +38,352 @@ init(State) ->
+        ?PRV_ERROR({failures_running_tests, _}) = Error ->
+            LogDir = proplists:get_value(logdir, CTOpts),
+            Index = filename:join([LogDir, "index.html"]),
+            ?CONSOLE("Results written to ~p.", [Index]),
```
Should we have that HTML notice in verbose results too? If the file is not generated, I'd like to see it in the help section.

talentdeficit replied:
yes probably. all the result handling stuff is from the old provider. i'm not sure why github considers it new code. i'll take a look at cleaning it up this afternoon

talentdeficit replied:
this is actually a non-problem. the ct output we suppress has the index url in it. we only print it when it wouldn't otherwise be printed

ferd replied:
Sounds good then

ferd commented:
Okay, I have a bunch comments, but otherwise it appears to be pretty cool. Spec file rewriting can be punted on until later.

marianoguerra commented:
hi! I have some questions regarding this change.

I have a plugin to do efene ct, https://github.com/efene/rebar3_efene_ct, what it does is to compile test/*.fn files to .beam files under the test folder and then invoke the ct task.

until now it didn't work since ct ignored my beam files, the only way to make it work was to compile them to .erl instead.

is this going to work with this change? 
is there a change I need to do to make it work?

talentdeficit commented:
@marianoguerra your provider should just work now, yes

talentdeficit commented:
i rebased this on current master and added some additional path handling to the shell and cover providers to work with the new test beam locations. no changes to `rebar_prv_common_test.erl`

ferd commented:
+1

marianoguerra commented:
I can confirm that efene quickstart now completely works against rebar3 master :)

http://efene.org/quickstart.html

PS: to test I built rebar3 from the repo instead of fetching from nightly, this will change when the next nightly is published of course.
